From c15851ce513435462c5d704b9b7c0369121af096 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:52 -0400
Subject: [ethernet] i40e: tolerate lost interrupts

Message-id: <1409658042-22689-232-git-send-email-sassmann@redhat.com>
Patchwork-id: 90432
O-Subject: [RHEL7.1 PATCH 231/281] i40e: tolerate lost interrupts
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

If the AQ interrupt gets lost for some reason, VF communications will
stall as the VFs have no way of reaching the PF, which is essentially
deaf. The VFs end up waiting forever for a reply that will never come.

To alleviate this condition, go ahead and check the ARQ every time we
run the service task. Remove the check for a pending event, and get rid
of a chatty error message that is now meaningless.

Change-ID: I0fc9d18169cd45c98f60188aef872cd6cee9a027
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Tested-by: Jim Young <jamesx.m.young@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 56497978bcbcde7a310ebaf2b67a936c66397593)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index f1cb17a..529ab3e 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -5314,9 +5314,6 @@ static void i40e_clean_adminq_subtask(struct i40e_pf *pf)
  u32 oldval;
  u32 val;
 
- if (!test_bit(__I40E_ADMINQ_EVENT_PENDING, &pf->state))
-  return;
-
  /* check for error indications */
  val = rd32(&pf->hw, pf->hw.aq.arq.len);
  oldval = val;
@@ -5360,10 +5357,9 @@ static void i40e_clean_adminq_subtask(struct i40e_pf *pf)
  do {
   event.msg_size = I40E_MAX_AQ_BUF_SIZE; /* reinit each time */
   ret = i40e_clean_arq_element(hw, &event, &pending);
-  if (ret == I40E_ERR_ADMIN_QUEUE_NO_WORK) {
-   dev_info(&pf->pdev->dev, "No ARQ event found\n");
+  if (ret == I40E_ERR_ADMIN_QUEUE_NO_WORK)
    break;
-  } else if (ret) {
+  else if (ret) {
    dev_info(&pf->pdev->dev, "ARQ event error %d\n", ret);
    break;
   }
-- 
1.7.1